### ISSUE TITLE
What is the purpose of IdenticalEqual mutator?

### ISSUE THREAD

**Issue by deleugpn** [infection/infection#438]
Consider the following snippet of code:

```
    private function runningOnTestEnvironment()
    {
        $repository = $this->app->get('config');

        return $repository->get('environment') === 'local';
    }
```

When running infection, I get a mutator that replaces `===` with `==` which cannot be killed. I know that it's a harmless mutation, but I would like to understand the reasoning behind this mutator.

For instance, the Function Signature mutators change methods from `public` to `protected` and `protected` to `private`, which means they help us make the public API of a class as strict as possible. In that sense, I would expect a mutation that changes `==` to `===`, but not the other way around.

Thanks for the attention and for the amazing tool!

**Comment by addiks** [comment]
This mutation can be killed. It disabled the type-strictness of the "===" operator. When mutated to the "==" operator, you can provoke invalid behavior by having "$repository->get('environment')" return a non-string that due to PHP's type-unsafety is "equal" to the compared string.

`
php > var_dump(0 === 'local');
bool(false)
php > var_dump(0 == 'local');
bool(true)
php >
`

**Comment by deleugpn** [comment]
Thanks for the insight, just a small follow up before I close this: Suppose the interface `public function get(string $key): string;` in a file that makes use of strict declaration. This means that `$repository->get()` can never return something that is not a string. In that case, should I just drop the `===` to `==` instead? 
My point being that given two method signatures that are strictly `string`, a strict comparison between them generates a mutation that cannot be killed, correct?

**Comment by BackEndTea** [comment]
> My point being that given two method signatures that are strictly string, a strict comparison between them generates a mutation that cannot be killed, correct?

This is correct, there are cases where this mutator leads to an 'unkillable' mutation. Unfortunately, we can't make the mutator 'smart' enough to detect those cases, but if you have any ideas how we could improve that we welcome a PR to fix that.

If you wish you could also turn off this mutator, (and the `not` version) like so:

```json
{
    "mutators": {
        "@default": true,
        "IdenticalEqual": false,
        "NotIdenticalNotEqual": false
    }
}
```

**Comment by borNfreee** [comment]
To be honest, we had (and still have) internal debates about `IdenticalEqual`/`NotIdenticalEqual` mutators.

As for me - it's not 'unkillable', this mutator shows you that `===` is not neccessary there and you can safely use `==` (keeping in mind your comment that `get()` _always_ returns `string`).

However, this mutator probably conflicts with tools like PHP-CS-Fixer/PHPStan that are set to force strict comparison everywhere (blindly).

I agree with @BackEndTea - you can just disable it. Also, I think we should add some notes to the documentation since this mutator is quite tricky.

**Comment by Ocramius** [comment]
Just for reference, `==` is not valid PHP in my projects, as it is disabled by CS and static analysis rules, so I'm not very worried about the mutation.

For my own stuff, it will likely be disabled, but most users jumping into mutation testing are likely to discover strict comparison issues anyway, so it should live on.

**Comment by acasademont** [comment]
Hi @BackEndTea why detecting the types to not perform this mutation should be impossible? Is it for performance reasons? (I'm not familiar with infection's internals, just curios as I stumbled upon this issue but I don't want to remove the mutator either, it can find legit bad code). Thanks!!

**Comment by sanmai** [comment]
Infection is not aware of types except for obvious cases of constants. That's the problem.

They were disabled in recent versions. If you wish to re-enable them, it's easy:

    "mutators": {
        "@default": true,
        "@equal": true,
        "@identical": true
    }

**Comment by acasademont** [comment]
thanks @sanmai :)
